controllers/metal3.io/preprovisioningimage_controller.go

```diff
@@ -110,6 +110,13 @@ func (r *PreprovisioningImageReconciler) Reconcile(ctx context.Context, req ctrl
 		log.Info("requeuing to check for secret", "after", delay)
 		result.RequeueAfter = delay
 	}
+
+	notReady := imageprovider.ImageNotReady{}
+	if errors.As(err, &notReady) {
```

Comment:
This isn't wrong and probably will work. But in such cases;
I prefer such usage;

```
if inr, ok := err.(imageprovider.ImageNotReady{}); ok {
				result.RequeueAfter = inr.RequeueAfter
			}
```
This looks more simpler to me, but as said your solution also looks good to me.

Reply:
Your idea looks better, but there is an interesting discussion in the overview of https://pkg.go.dev/errors. Do you think it applies?

Reply:
> In this case, it's not clear to me where such kind of error should be generated 

In the implementation. There is no guarantee that an image will be available quickly enough to return it immediately.

Reply:
It's unlikely that anyone would wrap this error (slightly more likely they'd use it to wrap something else), but even so I think it's preferable to use `errors.As()` just as a general policy.

It's kind of like the difference between `type(x) == Foo` vs. `isinstance(x, Foo)` in Python. (But not the same thing, obviously.)

Reply:
> > In this case, it's not clear to me where such kind of error should be generated
> 
> In the implementation. There is no guarantee that an image will be available quickly enough to return it immediately.

Just for the sake of clarity, do you mean by one of the methods of the `ImageProvider` interface? (In such case I'm not sure if we have some documentation for the expected errors, it could be worth adding it for making clearer the relationships between the reconciler and the image provider; ie `BuildImage` could return a `ImageBuildInvalid` error)